mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala

```diff
@@ -197,53 +196,43 @@ class StringIndexer @Since("1.4.0") (
     }
   }
 
-  private def countByValue(
-      dataset: Dataset[_],
-      inputCols: Array[String]): Array[OpenHashMap[String, Long]] = {
+  private def sortByFreq(dataset: Dataset[_], ascending: Boolean): Array[Array[String]] = {
+    val (inputCols, _) = getInOutCols()
     val selectedCols = getSelectedCols(dataset, inputCols.toImmutableArraySeq)
-    val results = Array.fill(selectedCols.size)(new OpenHashMap[String, Long]())
+    val numCols = inputCols.length
+
+    // In the case of equal frequency, always sorts strings by alphabet (ascending).
+    val countCol = if (ascending) count(lit(1)) else negate(count(lit(1)))
+
+    val result = Array.fill(numCols)(Array.empty[String])
     dataset.select(posexplode(array(selectedCols: _*)).as(Seq("index", "value")))
       .where(col("value").isNotNull)
       .groupBy("index", "value")
-      .agg(count(lit(1)).as("count"))
+      .agg(countCol.as("count"))
       .groupBy("index")
-      .agg(collect_list(struct("value", "count")))
+      .agg(collect_list(struct("count", "value")).alias("array"))
+      .repartition(8, col("index"))
```

Comment:
Out of curiosity, why 8 here? can we avoid the repartition()? I don't immediately see why we need it

Reply:
to be consistent with original parallelism:
```
    ThreadUtils.parmap(orgStrings, "sortingStringLabels", 8) { counts =>
      counts.toSeq.sortWith(sortFunc).map(_._1).toArray
    }.toArray
```

on second thought, I think we can remove it

Reply:
in the benchmark, removing it won't hurt the performance:
```
Elapsed time: 10.855109542s
```

Reply:
I get it, just seems like a repartition() is potentially expensive. I don't know if there was much significance to 8, other than it being a bit parallel. Here we already have parallelism in place and it's a small operation. I'd imagine you can remove it for simplicity, even if it isn't faster at small scale.